File: kotlin-mp-collections-jvm/src/main/java/edu/alex/ObjectReference.java
package edu.alex;

interface ObjectReference<T> {
    T get();

    @SuppressWarnings("unchecked")
    static <T> ObjectReference<T> wrap(final T object) {
        if(object == null) {
            return (ObjectReference) nullReference;
        }
        if(object instanceof String || object instanceof Integer || object instanceof Boolean) {
            return new PrimitiveReference(object);
        }

        return new NonPrimitiveReference(object);
    }

    ObjectReference<Object> nullReference = new ObjectReference<Object>() {
        @Override
        public Object get() {
            return null;
        }

        @Override
        public int hashCode() {
            return 0;
        }

        @Override
        public boolean equals(Object obj) {
            return obj == null || obj instanceof ObjectReference && ((ObjectReference) obj).get() == null;
        }
    };

    class PrimitiveReference<T> implements ObjectReference<T> {
        private final T ref;

        PrimitiveReference(final T ref) {
            this.ref = ref;
        }

        @Override
        public T get() {
            return this.ref;
        }

        @Override
        public int hashCode() {
            return this.ref.hashCode();
        }

        @Override
        public boolean equals(Object obj) {
            return obj instanceof ObjectReference && this.ref.equals( ((ObjectReference) obj).get() );
        }
    }

    class NonPrimitiveReference<T> implements ObjectReference<T> {
        private final T ref;

        NonPrimitiveReference(final T ref) {
            this.ref = ref;
        }

        @Override
        public T get() {
            return this.ref;
        }

        @Override
        public int hashCode() {
            return System.identityHashCode(this.ref);
        }

        @Override
        public boolean equals(Object obj) {
            return obj instanceof ObjectReference && this.ref == ((ObjectReference) obj).get();
        }
    }
}

File: src/main/jvm/edu/alex/ObjectReference.kt
package edu.alex

internal interface ObjectReference<T> {
    fun get(): T

    class PrimitiveReference<T> internal constructor(private val ref: T) : ObjectReference<T> {
        override fun get(): T {
            return this.ref
        }

        override fun hashCode(): Int {
            return this.ref.hashCode()
        }

        override fun equals(other: Any?): Boolean {
            return other is ObjectReference<*> && this.ref == other.get()
        }

        override fun toString() = "(Ref: ${this.ref})"
    }

    class NonPrimitiveReference<T> internal constructor(private val ref: T) : ObjectReference<T> {

        override fun get(): T {
            return this.ref
        }

        override fun hashCode(): Int {
            return System.identityHashCode(this.ref)
        }

        override fun equals(other: Any?): Boolean {
            return other is ObjectReference<*> && this.ref === other.get()
        }

        override fun toString() = "(Ref: ${this.ref})"
    }

    companion object {

        private fun isPrimitive(obj: Any) = obj is String || obj is Int || obj is Boolean || obj is Double

        @Suppress("UNCHECKED_CAST")
        fun <T> wrap(obj: T?): ObjectReference<T> {
            if (obj == null) {
                return nullReference as ObjectReference<T>
            }
            return if (isPrimitive(obj)) PrimitiveReference(obj) else NonPrimitiveReference(obj)
        }

        private val nullReference: ObjectReference<Any?> = object : ObjectReference<Any?> {
            override fun get(): Any? {
                return null
            }

            override fun hashCode(): Int {
                return 0
            }

            override fun equals(other: Any?): Boolean {
                return other == null || other is ObjectReference<*> && other.get() == null
            }

            override fun toString() = "(Ref: NULL)"
        }
    }
}